docs/book/common_code/cluster_controller.md

```diff
@@ -0,0 +1,156 @@
+	Status ClusterStatus `json:"status,omitempty"`
+}
+```
+{% endmethod %}
+
+{% method %}
+## ClusterSpec
+
+The `ClusterNetwork` field includes the information necessary to configure
+kubelet networking for `Pod`s and `Service`s.
+
+{% sample lang="go" %}
+```go
+// ClusterSpec defines the desired state of Cluster
+type ClusterSpec struct {
+	// Cluster network configuration
+	ClusterNetwork ClusterNetworkingConfig `json:"clusterNetwork"`
+
+	// Provider-specific serialized configuration to use during
+	// cluster creation. It is recommended that providers maintain
+	// their own versioned API types that should be
+	// serialized/deserialized from this field.
+	// +optional
+	ProviderSpec ProviderSpec `json:"providerSpec,omitempty"`
+}
+```
+{% endmethod %}
+
+{% method %}
+## ClusterStatus
+
+Some providers use the `APIEndpoint` field to determine when one or more
+masters have been provisioned. This may be necessary before worker nodes
+can be provisioned. For example, `cluster-api-provider-gcp` does [this](
+https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/f3145d8810a5c7fc434ddb5577699b4deb1b5fa6/pkg/cloud/google/metadata.go#L43):
+
+```go
+	if len(cluster.Status.APIEndpoints) == 0 {
+		return nil, fmt.Errorf("master endpoint not found in apiEndpoints for cluster %v", cluster)
+	}
+```
+
+**TODO**: Provide examples of how `ErrorReason` and `ErrorMessage` are 
+used in practice.
+
+{% sample lang="go" %}
+```go
+// ClusterStatus defines the observed state of Cluster
+type ClusterStatus struct {
+	// APIEndpoint represents the endpoint to communicate with the IP.
+	// +optional
+	APIEndpoints []APIEndpoint `json:"apiEndpoints,omitempty"`
+
+	// NB: Eventually we will redefine ErrorReason as ClusterStatusError once the
+	// following issue is fixed.
+	// https://github.com/kubernetes-incubator/apiserver-builder/issues/176
+
+	// If set, indicates that there is a problem reconciling the
+	// state, and will be set to a token value suitable for
+	// programmatic interpretation.
+	// +optional
+	ErrorReason common.ClusterStatusError `json:"errorReason,omitempty"`
+
+	// If set, indicates that there is a problem reconciling the
+	// state, and will be set to a descriptive error message.
+	// +optional
+	ErrorMessage string `json:"errorMessage,omitempty"`
+
+	// Provider-specific status.
+	// It is recommended that providers maintain their
+	// own versioned API types that should be
+	// serialized/deserialized from this field.
+	// +optional
+	ProviderStatus *runtime.RawExtension `json:"providerStatus,omitempty"`
+}
+```
+{% endmethod %}
+
+{% method %}
+## Cluster Actuator Interface
+
+All methods should be idempotent.
+
+`Reconcile()` will be called whenever there is a change to the `Cluster` 
+`Spec`, or after every resync period.
+
+If a `Cluster` resource is deleted, the controller will call the actuator's 
+`Delete()` method until it succeeds, or the finalizer is removed (see below).
```

Comment:
Should we link out to https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers on the word finalizer? alternatively we could make sure the CRD page is background reading including advanced topics